src/Dialect/ONNX/ONNXOps.cpp

```diff
@@ -597,6 +597,46 @@ static LogicalResult RNNShapeInference(T *op, int gates) {
   return success();
 }
 
+static void insertConvTransposePads(SmallVectorImpl<int64_t> &inferedPads,
+    const llvm::StringRef autoPad, ArrayRef<int64_t> xShape,
+    Optional<ArrayAttr> kernelShape, Optional<ArrayAttr> padsOpt,
```

Comment:
1 more thing: since the code below assumes kernelShape, stridesOpt, outputShapeOpt all satisfy has_value() (otherwise they hit assertion failure in ArrayAttrIntVal()), I think it will be clearer if they are passed into insertConvTransposePads with type ArrayAttr and change the caller to call insertConvTransposePads with .value() of all those args

Reply:
or forget about it if that doesn't match the existing logic for insertConvTransposeSpatialDim

Reply:
Yeah, just copying the patterns in the file.

I'll leave a refactor for things like this for another PR

Reply:
how different is this from the regular conv? Also, could it be packaged into a shape inference helper class... trying to reduce the amount of specific shape inference inside ONNXOps.cpp...